### ISSUE TITLE
State of the art regression/classification

### ISSUE THREAD

**Issue by jmmcd** [jmmcd/PonyGE2#27]
For near-state of the art performance on regression and classification, I think we need:

* Gradient descent on constants, i.e. the grammar creates a constant called C0, C1, etc rather than a literal constant, and then we use a gradient descent method to optimise the constants per-individual, with CMA-ES, BFGS, or similar

* Multi-objective optimisation using NSGA2 or similar

* Auto-simplification using Sympy or similar.

I have code for these, I will try to add it eventually.

**Comment by jmmcd** [comment]
Other regression/classification tasks:

* unify the regression and classification code

* add code for adding the right number of independent variables to the grammar, using code like this:

```python
<var> ::= x[<varidx>]
<varidx> ::= GERANGE:n_vars
```
where `n_vars` is automatically derived from the size of the training set.

**Comment by jmmcd** [comment]
Done:

* optimisation of constants
* unify regression/classification
* generic grammars using `x[<varidx>]`

Remaining:

* Auto-simplification
* MOOOOOO

**Comment by dvpfagan** [comment]
I assume auto simplification will only happen at end of run?

**Comment by jmmcd** [comment]
Yeah I think so.

**Comment by jmmcd** [comment]
I'm thinking of changing the implementation here.

When individuals start getting big, we can easily have 50-100 constants. Optimisation of that many becomes very slow. And we probably don't always need so many. So I can make it so that the grammar creates a limited number of constants (maybe the same as number of variables?), but is allowed to use each of them more than once.

**Comment by jmmcd** [comment]
Done:
* Use grammar to say how many constants you want
* Refactoring to a new utils/fitness/optimize_constants.py file.

**Comment by mikefenton** [comment]
NSGA-II now mostly implemented.

**Comment by mikefenton** [comment]
What is left to do on this? Is it just simplification, or is there more?

**Comment by jmmcd** [comment]
Just the simplification.
